src/library/congr_lemma.cpp

```diff
@@ -363,15 +363,38 @@ struct congr_lemma_manager {
         to_buffer(finfo.get_params_info(), pinfos);
         to_buffer(ssinfos, ssinfos_buffer);
         kinds.resize(pinfos.size(), congr_arg_kind::Eq);
+
+        // The default congr_arg_kind is Eq, which allows `simp` to rewrite this
+        // argument. However, if there are references from `i` to `j`, we cannot
+        // rewrite both `i` and `j`. So we must change the congr_arg_kind at
+        // either `i` or `j`. In principle, if there is a dependency with `i`
+        // appearing after `j`, then we set `j` to Fixed (or Cast). But there is
+        // an optimization: if `i` is a subsingleton, we can fix it instead of
+        // `j`, since all subsingletons are equal anyway. The fixing happens in
+        // two loops: one for the special cases, and one for the general case.
         for (unsigned i = 0; i < pinfos.size(); i++) {
             if (std::find(result_deps.begin(), result_deps.end(), i) != result_deps.end()) {
                 kinds[i] = congr_arg_kind::Fixed;
             } else if (ssinfos_buffer[i].is_subsingleton()) {
-                // See comment at mk_congr.
-                if (!pinfos[i].is_prop() && pinfos[i].has_fwd_deps())
-                    kinds[i] = congr_arg_kind::Fixed;
-                else
+                if (pinfos[i].is_prop()) {
+                    // Propositions are all definitionally equal, so we don't need to make this Eq.
                     kinds[i] = congr_arg_kind::Cast;
+                } else if (pinfos[i].has_fwd_deps()) {
+                    // If something depends on `i`, it should be Fixed.
+                    kinds[i] = congr_arg_kind::Fixed;
+                } else {
+                    // It's not a Prop and it has no forwards dependencies.
+                    // If there are backwards dependencies on Eq arguments, then
+                    // we will use subsingleton elimination to prove this
+                    // equality.
+                    // Otherwise we can let this arg stay Eq without problems.
```

Comment:
This introduces a counterintuitive change for instance-implicit arguments without backward dependencies: simp can rewrite the argument if and only if it is a subsingleton.

I think it's a bad idea to make the decision where we can rewrite based on whether there is a subsingleton instance available.  That's what's got us in this situation in the first place.

I think the conservative approach would be to always mark instance-implicit arguments as Fixed (or Cast, if possible).